### PR TITLE
[SPARK-23627][SQL] Provide isEmpty in Dataset

### PR DESCRIPTION
**Pull request by goungoun**
## What changes were proposed in this pull request?

This PR adds isEmpty() in DataSet

## How was this patch tested?

Unit tests added

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -511,6 +511,12 @@ class Dataset[T] private[sql](
    */
   def isLocal: Boolean = logicalPlan.isInstanceOf[LocalRelation]
 
+  /**
+   * Returns true if the `DataSet` is empty
+   *
```
Comment (mgaido91) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: this would need a 
```
@since 2.4.0
```

Comment (goungoun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @mgaido91 Thanks, I modified the comment.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -511,6 +511,14 @@ class Dataset[T] private[sql](
    */
   def isLocal: Boolean = logicalPlan.isInstanceOf[LocalRelation]
 
+  /**
+   * Returns true if the `DataSet` is empty
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Dataset

**Comment by mgaido91** [comment]
LGTM

**Comment by goungoun** [comment]
@HyukjinKwon, @maropu 
Just a gentle reminder. Jenkins is waiting for a comment like 'ok to test'.

**Comment by maropu** [comment]
LGTM, too.

**Comment by rxin** [comment]
So the API looks useful, but I don't know if this is the right implementation. How important is it to add this? It seems like the value is not super high either.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -511,6 +511,14 @@ class Dataset[T] private[sql](
    */
   def isLocal: Boolean = logicalPlan.isInstanceOf[LocalRelation]
 
+  /**
+   * Returns true if the `Dataset` is empty
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: (dot at the end)

Comment (goungoun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @HyukjinKwon, thanks. I added period.

**Comment by goungoun** [comment]
@rxin, checking empty is likely to be a common process in every ETL batch job. I think it is the right place to provide that functionality. When a basic function is missing already supposed to be provided, people spend unnecessary time for searching and creating their own creative functions. It does not help us develop clean code or business value neither. I added one of the stack-overflow discussions at SPARK-23627 for your reference. I also would like to confirm rdd.isEmpty is optimized internally following up this issue.

**Comment by goungoun** [comment]
For additional check that I mentioned. The following code shows that Spark users does not need to add take(1). ds.rdd.take(1).isEmpty is redundant.

[RDD.scala](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala) 
`def isEmpty(): Boolean = withScope {
    partitions.length == 0 || take(1).length == 0
}`

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -511,6 +511,14 @@ class Dataset[T] private[sql](
    */
   def isLocal: Boolean = logicalPlan.isInstanceOf[LocalRelation]
 
+  /**
+   * Returns true if the `Dataset` is empty.
+   *
+   * @group basic
+   * @since 2.4.0
+   */
+  def isEmpty: Boolean = rdd.isEmpty()
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Building a rdd is not cheap. The current impl does not perform well.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: How about this?
```
  def isEmpty: Boolean = withAction("isEmpty", groupBy().count().queryExecution) { plan =>
    plan.executeCollect().head.getLong(0) == 0
  }
```

Comment (goungoun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @gatorsmile, just simply running df.rdd.isEmpty in spark-shell was quite responsive even in tera byte sized tables.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @goungoun maybe; it depends on the Dataset. I'd follow @maropu 's lead or close this.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: If it's worth doing, I'll take (if nobody does this...)

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `RDD#isEmpty` is pretty effective, it just checks if all the partitions are empty, without loading the data. The problem is how to build an RDD from `Dataset`, which minimize the cost of building the `Iterator`.

It seems `Dataset#rdd` is not good enough, e.g., if we have a `Filter` in the query, we may do a full scan(no column pruning) for the underlying files.

Doing a count is not perfect either. Ideally we can stop as soon as we see one record.

I'd suggest doing a `limit 1` first and then count.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: +1 for the limit

Comment (goungoun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: In my understanding, the @maropu's suggestion is to implement isEmpty exactly the same way of count() method except the return type. 

```
def count(): Long = withAction("count", groupBy().count().queryExecution) { plan =>
   plan.executeCollect().head.getLong(0)
 }

def isEmpty: Boolean = withAction("isEmpty", groupBy().count().queryExecution) { plan =>
  plan.executeCollect().head.getLong(0) == 0
}
```
And @cloud-fan 's idea is adding limit(1) before doing count! Is it right?

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**